lib/Sema/SemaBounds.cpp

```diff
@@ -628,93 +698,210 @@ namespace {
           }
+          // `e - p` has empty bounds, regardless of the bounds of p.
+          // `e -= p` has empty bounds, regardless of the bounds of p.
+          if (RHS->getType()->isPointerType() &&
+              Op == BinaryOperatorKind::BO_Sub) {
+            return CreateBoundsEmpty();
+          }
 
           // Arithmetic on integers with bounds.
+          //
+          // `e1 @ e2` has the bounds of whichever of `e1` or `e2` has bounds.
+          // if both `e1` and `e2` have bounds, then they must be equal.
+          // Both `e1` and `e2` are RValues
+          //
+          // `e1 @= e2` has the bounds of whichever of `e1` or `e2` has bounds.
+          // if both `e1` and `e2` have bounds, then they must be equal.
+          // `e1` is an LValue, its bounds are the lvalue target bounds.
+          // `e2` is an RValue
+          //
+          // @ can stand for: +, -, *, /, %, &, |, ^, >>, <<
           if (LHS->getType()->isIntegerType() &&
               RHS->getType()->isIntegerType() &&
               (BinaryOperator::isAdditiveOp(Op) ||
                BinaryOperator::isMultiplicativeOp(Op) ||
-               BinaryOperator::isBitwiseOp(Op))) {
+               BinaryOperator::isBitwiseOp(Op) ||
+               BinaryOperator::isShiftOp(Op))) {
             BoundsExpr *LHSBounds = IsCompoundAssignment ?
               LValueTargetBounds(LHS) : RValueBounds(LHS);
             BoundsExpr *RHSBounds = RValueBounds(RHS);
             if (LHSBounds->isNone() && !RHSBounds->isNone())
               return RHSBounds;
             if (!LHSBounds->isNone() && RHSBounds->isNone())
               return LHSBounds;
-            // TODO: check if both LHS and RHS have bounds and
-            // indicate cause of failure to have bounds.
+            if (!LHSBounds->isNone() && !RHSBounds->isNone()) {
+              // TODO: Check if LHSBounds and RHSBounds are equal.
+              // if so, return one of them. If not, return bounds(none)
+              return CreateBoundsUnknown();
+            }
+            if (LHSBounds->isNone() && RHSBounds->isNone())
+              return CreateBoundsEmpty();
           }
-          return CreateBoundsNone();
+
+          // Comparisons and Logical Ops
+          //
+          // `e1 @ e2` have empty bounds if @ is:
+          // ==, !=, <=, <, >=, >, &&, ||
+          if (BinaryOperator::isComparisonOp(Op) ||
+              BinaryOperator::isLogicalOp(Op)) {
+            return CreateBoundsEmpty();
+          }
+
+          // All Other Binary Operators we don't know how to deal with
+          return CreateBoundsEmpty();
+        }
+        case Expr::CallExprClass: {
+          const CallExpr *CE = dyn_cast<CallExpr>(E);
+          if (!CE) {
+            llvm_unreachable("unexpected cast failure");
+            return CreateBoundsInferenceError();
+          }
+
+          // Get function prototype.
+          // The Callee is a function *pointer*, almost always.
```

Comment:
I should revisit this comment, the callee is always a function pointer, hence needing to `getPointeeType()`